### ISSUE TITLE
Error: Cannot find module '...../bcrypt/lib/binding/napi-v3/bcrypt_lib.node'

### ISSUE THREAD

**Issue by jellohouse** [kelektiv/node.bcrypt.js#800]
After installing the newest version of node 12.6.3 and installing bcrypt 4.0.1 in my project,
"npm install bcrypt" added the package successfully without errors, but then running "node app.js" I get:

`Error: Cannot find module 'path_to_my_project/node_modules/bcrypt/lib/binding/napi-v3/bcrypt_lib.node'`

And indeed the folder `node_modules/bcrypt/lib` is not there at all...

Why would installation not make this folder?
I suspect it's something to do with node-gyp not doing it's thing, but why don't I get an error message when i do npm install?

And I tried reverting to my old node version 8.11.1 using https://www.npmjs.com/package/n to manage my node versions and then installing bcrypt 2.0.1 (like I had previously), but it's behaving the exact same way... npm install adds the package but the bcrypt/lib folder is missing.
I also tried with node 12.6.1 and 12.6.2, same thing.

**Comment by jellohouse** [comment]
Ok for anyone else having this issue here is how I solved it, I had to compile it manually:
`cd /node_modules/bcrypt`
`npm install requests` (https://github.com/kelektiv/node.bcrypt.js/issues/796#issuecomment-627982852)
`node-pre-gyp install --fallback-to-build`

**Comment by spicecoder** [comment]
> node-pre-gyp install --fallback-to-build.. solved my issue too thanks

**Comment by joy45689** [comment]
> Ok for anyone else having this issue here is how I solved it, I had to compile it manually for some reason:
> `cd node_modules/bcrypt`
> `node-pre-gyp install --fallback-to-build`

It solves my problem too. Thanks.

**Comment by vishalims095** [comment]
sudo npm i --unsafe-perm
after that install bcrypt


its works for me ubuntu 18.04

**Comment by lehathanh** [comment]
> Ok for anyone else having this issue here is how I solved it, I had to compile it manually for some reason:
> `cd node_modules/bcrypt`
> `node-pre-gyp install --fallback-to-build`

This works for me on newly installation on Windows 10/VSC, nodeJS v14.15.1. Thanks a lot!

**Comment by loveyousomuch554** [comment]
> Ok for anyone else having this issue here is how I solved it, I had to compile it manually for some reason:
> `cd node_modules/bcrypt`
> `node-pre-gyp install --fallback-to-build`

It helps me. Thanks

**Comment by AITuring** [comment]
if mac still have problems after `node-pre-gyp install --fallback-to-build`, maybe it's mac [problem](https://stackoverflow.com/questions/52522565/git-is-not-working-after-macos-update-xcrun-error-invalid-active-developer-pa). You should fix xcun.

**Comment by Shishoo** [comment]
node version: **14.16.0 (LTS)**
npm version:  **6.14.11**
OS: **Windows 10**

I get the following error:

```javascript
internal/modules/cjs/loader.js:883
  throw err;
  ^
 
Error: Cannot find module '...path...\node_modules\bcrypt\lib\binding\napi-v3\bcrypt_lib.node'
Require stack:
- ...path...\node_modules\bcrypt\bcrypt.js
- ...path...\crypto.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (...path...\node_modules\bcrypt\bcrypt.js:6:16)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '...path...\\node_modules\\bcrypt\\bcrypt.js',
    '...path...\\crypto.js'
  ]
}
```

I have tried the fixes here and [here](https://github.com/kelektiv/node.bcrypt.js/issues/800) and haven't been able to resolve the issue.

**Comment by recrsn** [comment]
@Shishoo it appears that your bcrypt DLL is missing. Run `npm rebuild bcrypt`

**Comment by Shishoo** [comment]
@recrsn 
Hi Amitosh - with reference to your [previous post here](https://github.com/kelektiv/node.bcrypt.js/issues/566), I did try 
`npm rebuild bcrypt` - unfortunately this didn't work.

**Comment by recrsn** [comment]
What's the error while trying to rebuild? 

Do you have an outer layer of package.json?

**Comment by Shishoo** [comment]
Here's everything I've encountered and have.

```javascript
----------------------------------------------------
> npm i bcrypt
----------------------------------------------------

npm WARN deprecated node-pre-gyp@0.15.0: Please upgrade to @mapbox/node-pre-gyp: the non-scoped node-pre-gyp package is deprecated and only the @mapbox scoped package will recieve updates in the future
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN bcrypto@1.0.0 No description
npm WARN bcrypto@1.0.0 No repository field.

+ bcrypt@5.0.0
added 68 packages from 116 contributors and audited 68 packages in 13.075s

1 package is looking for funding
  run `npm fund` for details    

found 0 vulnerabilities

--------------------------------------------------
> node app.js
----------------------------------------------------
internal/modules/cjs/loader.js:883
  throw err;
  ^

Error: Cannot find module 'F:\...path-to-file\node_modules\bcrypt\lib\binding\napi-v3\bcrypt_lib.node'
Require stack:
- F:\Learning\__Tuts\ColtSteele\S50-492-Auth\bcrypto\node_modules\bcrypt\bcrypt.js
- F:\Learning\__Tuts\ColtSteele\S50-492-Auth\bcrypto\app.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (F:\...path-to-file\node_modules\bcrypt\bcrypt.js:6:16)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'F:\\...path-to-file\\node_modules\\bcrypt\\bcrypt.js',
    'F:\\...path-to-file\\bcrypto\\app.js'
  ]
}

----------------------------------------------------

> npm rebuild bcrypt
bcrypt@5.0.0 F:\...path-to-file\bcrypto\node_modules\bcrypt

(No errors or warnings)

----------------------------------------------------
> node app.js
----------------------------------------------------
internal/modules/cjs/loader.js:883
  throw err;
  ^

Error: Cannot find module 'F:\...path-to-file\node_modules\bcrypt\lib\binding\napi-v3\bcrypt_lib.node'
Require stack:
- F:\Learning\__Tuts\ColtSteele\S50-492-Auth\bcrypto\node_modules\bcrypt\bcrypt.js
- F:\Learning\__Tuts\ColtSteele\S50-492-Auth\bcrypto\app.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (F:\...path-to-file\node_modules\bcrypt\bcrypt.js:6:16)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'F:\\...path-to-file\\node_modules\\bcrypt\\bcrypt.js',
    'F:\\...path-to-file\\bcrypto\\app.js'
  ]
}

----------------------------------------------------
contents of: package.json
----------------------------------------------------
{
  "name": "cryptoapplibtests",
  "version": "1.0.0",
  "description": "testing bcrypt",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "bcrypt": "^5.0.0"
  }
}

----------------------------------------------------
contents of: app.js
----------------------------------------------------

// REF: https://www.npmjs.com/package/bcrypt
const bcrypt = require('bcrypt');

pwd = '£MonkeyMa9ic';

const hashPassword = async (pw) => {
  const salt = await bcrypt.genSalt(11);
  const hash = await bcrypt.hash(pw, salt)

  console.log(salt);
  console.log(hash);
};
```

**Comment by manoharreddyporeddy** [comment]
Below solved the issue:

1. // delete "node_modules"  OR rename "node_modules" to something else
2. `npm cache clean --force`
3. `npm i`

**Comment by sk8Guerra** [comment]
> @Shishoo it appears that your bcrypt DLL is missing. Run `npm rebuild bcrypt`

Thanks man! That helped solve the error.

**Comment by XiaoyunWu6666** [comment]
> Ok for anyone else having this issue here is how I solved it, I had to compile it manually for some reason:
> `cd node_modules/bcrypt`
> `node-pre-gyp install --fallback-to-build`

The only answer online that eventually helped me :)

**Comment by jongomes** [comment]
not work me

**Comment by mgalvao2012** [comment]
I fixed this issue this way. On the terminal from root folder of the project:
cd node_modules/bcrypt
make

**Comment by duttaANI** [comment]
> I fixed this issue this way. On the terminal from root folder of the project: cd node_modules/bcrypt make

This solved the issue. Thanks a lot.

**Comment by AhmedHajeyah** [comment]
> I fixed this issue this way. On the terminal from root folder of the project: cd node_modules/bcrypt make

Solved my issue

**Comment by axel-lukongo** [comment]
> sudo npm i --unsafe-permaprès cela, installez bcrypt
> 
> ça marche pour moi ubuntu 18.04

ça à aussi fonctionner pour moi et je suis sur Mac os

**Comment by Samuelfapohunda** [comment]
> @Shishoo it appears that your bcrypt DLL is missing. Run `npm rebuild bcrypt`

This worked for me

**Comment by mostafaSalah12** [comment]
> Ok for anyone else having this issue here is how I solved it, I had to compile it manually for some reason: `cd node_modules/bcrypt` `node-pre-gyp install --fallback-to-build`

This solves my problem. Thanks.

**Comment by SachinJadeja** [comment]
This error is coming up again in Bun(NextJS 13.5.6, Node 20.7.0) This error is coming up again in Bun(NextJS 13.5.6, Node 20.7.0)

**Comment by maztak** [comment]
i use pnpm and node.js@18.16.0.
i have to use pnpm@8.3. pnpm@8.10 deesn't work.

**Comment by cookabc** [comment]
> node-pre-gyp install --fallback-to-build

Save my day, thank you sir!

**Comment by cookabc** [comment]
> I fixed this issue this way. On the terminal from root folder of the project: cd node_modules/bcrypt make

save my day, thank you sir!

**Comment by codebyBowen** [comment]
> Ok for anyone else having this issue here is how I solved it, I had to compile it manually for some reason: `cd node_modules/bcrypt` `node-pre-gyp install --fallback-to-build`

Big thanks, works for me

**Comment by Gxmadix** [comment]
I was missing `node-gyp` on my system, after installing it and rebuilding like told by @recrsn `npm rebuild bcrypt ` this solved my issue

**Comment by lutfi-haslab** [comment]
this work for me, for macbook air m1
https://github.com/kelektiv/node.bcrypt.js/issues/800#issuecomment-785182628

**Comment by kurumeii** [comment]
@kireerik
i still have this issue, but i make it work by deleting the **node_modules** folder, then `pnpm install `it again and it worked

**Comment by eldardada** [comment]
> > Ok for anyone else having this issue here is how I solved it, I had to compile it manually for some reason:
> > `cd node_modules/bcrypt`
> > `node-pre-gyp install --fallback-to-build`
> 
> It solves my problem too. Thanks.



> I was missing `node-gyp` on my system, after installing it and rebuilding like told by [@recrsn](https://github.com/recrsn) `npm rebuild bcrypt ` this solved my issue

Thanks, it works. I think this is the best solution because I don't have to go into node_modules/bcrypt every time and install something manually

**Comment by cbelsole** [comment]
If you're using pnpm v10 it prevents packages from running install scripts. [This comment goes into more detail.](https://github.com/pnpm/pnpm/issues/9042#issuecomment-2638094544)

**Comment by erenjs** [comment]
> I was missing `node-gyp` on my system, after installing it and rebuilding like told by [@recrsn](https://github.com/recrsn) `npm rebuild bcrypt ` this solved my issue

After rebuilding bcrypt I got this error:

> Error: Something went wrong installing the "sharp" module Cannot find module '../build/Release/sharp-win32-x64.node'

But after rebuilding sharp the application started successfully:
`npm rebuild sharp`

**Comment by JasonLamv-t** [comment]
If you are using pnpm, this issue might be caused by pnpm's update, check https://github.com/pnpm/pnpm/issues/9032.

The solution is adding this to your `package.json` file:

```json
  "pnpm": {
    "onlyBuiltDependencies": ["bcrypt"]
  },
```

**Comment by oendnsk675** [comment]
> If you're using pnpm v10 it prevents packages from running install scripts. [This comment goes into more detail.](https://github.com/pnpm/pnpm/issues/9042#issuecomment-2638094544)

thanks for the reference

**Comment by Jaan19867** [comment]
> Ok for anyone else having this issue here is how I solved it, I had to compile it manually for some reason: `cd node_modules/bcrypt` `node-pre-gyp install --fallback-to-build`

Thank you , It solved my issue. Spend 3 hr to debug it , and then came to your comment , 
Thanks again  .

**Comment by guscsales** [comment]
> [@eldardada](https://github.com/eldardada) [@erenjs](https://github.com/erenjs) If you are using pnpm, this issue might be caused by pnpm's update, check [pnpm/pnpm#9032](https://github.com/pnpm/pnpm/issues/9032).
> 
> The solution is adding this to your `package.json` file:
> 
>   "pnpm": {
>     "onlyBuiltDependencies": ["bcrypt"]
>   },

This works well, thanks!

**Comment by malikEhsan09** [comment]
> Ok for anyone else having this issue here is how I solved it, I had to compile it manually for some reason: `cd node_modules/bcrypt` `node-pre-gyp install --fallback-to-build`

This works for me after changing the node version from 20 to 18, nodeJS v18.20.2. Thanks a lot!

**Comment by kadirmetin** [comment]
I had the same problem while updating my project and I solved it by using “bcryptjs” instead of “bcrypt”.

**Comment by wintsa123** [comment]
node version 23

i use 
```js
yarn install i --unsafe-perm 
```
fix it

**Comment by chandeldivyam** [comment]
> [@eldardada](https://github.com/eldardada) [@erenjs](https://github.com/erenjs) If you are using pnpm, this issue might be caused by pnpm's update, check [pnpm/pnpm#9032](https://github.com/pnpm/pnpm/issues/9032).
> 
> The solution is adding this to your `package.json` file:
> 
>   "pnpm": {
>     "onlyBuiltDependencies": ["bcrypt"]
>   },

Thanks!

**Comment by h3cksamrat** [comment]
> [@eldardada](https://github.com/eldardada) [@erenjs](https://github.com/erenjs) If you are using pnpm, this issue might be caused by pnpm's update, check [pnpm/pnpm#9032](https://github.com/pnpm/pnpm/issues/9032).
> 
> The solution is adding this to your `package.json` file:
> 
>   "pnpm": {
>     "onlyBuiltDependencies": ["bcrypt"]
>   },

Thanks, this worked.

**Comment by ramanujamgond** [comment]
> [@eldardada](https://github.com/eldardada) [@erenjs](https://github.com/erenjs) If you are using pnpm, this issue might be caused by pnpm's update, check [pnpm/pnpm#9032](https://github.com/pnpm/pnpm/issues/9032).
> 
> The solution is adding this to your `package.json` file:
> 
>   "pnpm": {
>     "onlyBuiltDependencies": ["bcrypt"]
>   },

Thank you @JasonLamv-t - this method worked for me.

**Comment by cern7** [comment]
> [@Shishoo](https://github.com/Shishoo) it appears that your bcrypt DLL is missing. Run `npm rebuild bcrypt`

This worked for me :)

**Comment by MCesarczyk** [comment]
> "pnpm": {
>     "onlyBuiltDependencies": ["bcrypt"]
>   },

Thanks, in my case this worked when used with addition of 
```
RUN pnpm install
RUN pnpm rebuild bcrypt
```
in Dockerfile 🚀

**Comment by aozoragh** [comment]
Thanks, I have same issue with pnpm for some modules such like bcrypt, @nestjs/core, @apollo/protobufjs as well.
"pnpm" : {
     "onlyBuiltDependencies": [
           "@apollo/protobufjs",
           "@nestjs/core",
           "bcrypt"
      ]
}
makes me happy 😊

**Comment by arm5472** [comment]
I am using "pnpm", and tried  "pnpm rebuild bcrypt" with no luck. 

But this sorted the issue for me:

"npm rebuild bcrypt"

**Comment by recrsn** [comment]
pnpm v10+ users, please add `bcrypt` to `onlyBuiltDependencies` in your `package.json`, else pnpm will ignore building the native dependencies

**Comment by DanielReyes03** [comment]
> I am using "pnpm", and tried "pnpm rebuild bcrypt" with no luck.
> 
> But npm sorted the issue for me:
> 
> "npm rebuild bcrypt"

Hello everyone, I'm working with NestJS and pnpm. Everything ran smoothly during development, but when I tried deploying on Ubuntu (Azure) using Jenkins and Docker, I encountered issues with pnpm. I tested many solutions, but none worked for me. The only thing that solved the problem was switching to npm. After hours of troubleshooting, npm finally did the job.

For NestJS, I used these two lines in my Dockerfile:

`RUN npm ci --legacy-peer-deps --no-audit --progress=false`
`RUN npm rebuild bcrypt`

¡Happy Coding!
